R/bias.R

```diff
@@ -118,121 +112,126 @@ bias_sample <- function(true_values, predictions) {
-#'   4341.500, 4006.250, 1127.000, 705.500
-#' )
-#'
-#' upper <- c(
-#'   6341.000, 6352.500, 6594.986, 6978.500,
-#'   7231.000, 7341.500, 7860.004, 7973.000,
-#'   8340.500, 8675.750, 11555.000, 11976.500
+#' predictions <- c(
+#'   705.500, 1127.000, 4006.250, 4341.500, 4709.000, 4821.996,
+#'   5340.500, 5451.000, 5703.500, 6087.014, 6329.500, 6341.000,
+#'   6352.500, 6594.986, 6978.500, 7231.000, 7341.500, 7860.004,
+#'   7973.000, 8340.500, 8675.750, 11555.000, 11976.500
 #' )
 #'
-#' range <- c(0, 10, 20, 30, 40, 50, 60, 70, 80, 90, 95, 98)
+#' quantiles <- c(0.01, 0.025, seq(0.05, 0.95, 0.05), 0.975, 0.99)
 #'
 #' true_value <- 8062
 #'
-#' bias_range(
-#'   lower = lower, upper = upper,
-#'   range = range, true_value = true_value
-#' )
+#' bias_quantile(predictions, quantiles, true_value = true_value)
 #' @export
 #' @keywords metric
 
-bias_range <- function(range, lower, upper,
-                       true_value) {
-  lower_predictions <- lower
-  upper_predictions <- upper
+bias_quantile <- function(predictions, quantiles, true_value) {
+  # check that predictions and quantiles have the same length
+  if (!length(predictions) == length(quantiles)) {
+    stop("predictions and quantiles must have the same length")
+  }
+
+  if (anyNA(predictions)) {
+    quantiles <- quantiles[!is.na(predictions)]
+    predictions <- predictions[!is.na(predictions)]
+  }
+
+  if (anyNA(quantiles)) {
+    quantiles <- quantiles[!is.na(quantiles)]
+    predictions <- predictions[!is.na(quantiles)]
+  }
 
-  if (anyNA(upper) || anyNA(lower)) {
-    range <- range[!is.na(upper) & !is.na(lower)]
-    lower_predictions <- lower[!is.na(lower) & !is.na(upper)]
-    upper_predictions <- upper[!is.na(lower) & !is.na(upper)]
+  # if there is no input, return NA
+  if (length(quantiles) == 0 || length(predictions) == 0) {
+    return(NA_real_)
+  }
 
-    # deal with the point forecast case where inputs may be NA
-    if (length(range) == 0 ||
-        length(lower_predictions) == 0 ||
-        length(upper_predictions) == 0
-      ) {
-      return(NA_real_)
-    }
+  # quantiles must be between 0 and 1, increase, and be unique
+  if (any(quantiles < 0) || any(quantiles > 1)) {
+    stop("quantiles must be between 0 and 1")
   }
 
-  # convert range to quantiles
-  lower_quantiles <- abs(100 - range) / (2 * 100)
-  upper_quantiles <- abs(100 + range) / (2 * 100)
+  if (!all(diff(quantiles) > 0)) {
+    stop("quantiles must be increasing")
+  }
 
-  # handling if median is not available - compute median as mean between two
-  # inner most quantiles
-  if (0 %in% range) {
-    median_prediction <- upper_predictions[range == 0]
+  if (0.5 %in% quantiles) {
+    median_prediction <- predictions[quantiles == 0.5]
   } else {
+    # if median is not available, compute as mean of two innermost quantiles
+    message(
+      "Median not available, computing as mean of two innermost quantiles",
+      " in order to compute bias."
+    )
     median_prediction <-
```

Comment:
Should this be weighted mean / linear approximation if the innermost quantiles not equidistant from 0.5?

Reply:
Yes I agree. This is not a new change however so I think this is a new feature issue. I put the message in here to make it clear this is happening as before it happened silently.